hushline/premium.py

```diff
@@ -0,0 +1,609 @@
+
+            except Exception as e:
+                current_app.logger.error(
+                    f"Error processing event {stripe_event.event_type} ({stripe_event.event_id}): {e}\n{stripe_event.event_data}"  # noqa: E501
+                )
+                stripe_event.status = StripeEventStatusEnum.ERROR
+                stripe_event.error_message = str(e)
+                db.session.add(stripe_event)
+                db.session.commit()
+                continue
+
+            stripe_event.status = StripeEventStatusEnum.FINISHED
+            db.session.add(stripe_event)
+            db.session.commit()
+
+
+def create_blueprint(app: Flask) -> Blueprint:
+    # Now define the blueprint
+    bp = Blueprint("premium", __file__, url_prefix="/premium")
+
+    @bp.route("/", methods=["GET"])
+    @authentication_required
+    def index() -> Response | str:
+        user = db.session.get(User, session.get("user_id"))
+        if not user:
+            session.clear()
+            return redirect(url_for("login"))
+
+        # Check if we have an incomplete subscription
+        stripe_subscription = get_subscription(user)
+        if stripe_subscription and stripe_subscription["status"] == "incomplete":
+            flash("⚠️ Your subscription is incomplete. Please try again.", "warning")
+
+        # Load the user's invoices
+        invoices = db.session.scalars(
+            db.select(StripeInvoice)
+            .filter_by(user_id=user.id)
+            .filter_by(status=StripeInvoiceStatusEnum.PAID)
+            .order_by(desc(StripeInvoice.created_at))
+            .all()
+        )
+
+        return render_template(
+            "premium.html", user=user, invoices=invoices, business_price=get_business_price_string()
+        )
+
+    @bp.route("/select-tier", methods=["GET"])
+    @authentication_required
+    def select_tier() -> Response | str:
+        user = db.session.get(User, session.get("user_id"))
+        if not user:
+            session.clear()
+            return redirect(url_for("login"))
+
+        return render_template(
+            "premium-select-tier.html", user=user, business_price=get_business_price_string()
+        )
+
+    @bp.route("/select-tier/free", methods=["POST"])
+    @authentication_required
+    def select_free() -> Response | str:
+        user = db.session.get(User, session.get("user_id"))
+        if not user:
+            session.clear()
+            return redirect(url_for("login"))
+
+        if user.tier_id is None:
+            user.tier_id = FREE_TIER
+            db.session.add(user)
+            db.session.commit()
+
+        return redirect(url_for("inbox"))
+
+    @bp.route("/waiting", methods=["GET"])
+    @authentication_required
+    def waiting() -> Response | str:
+        return render_template("premium-waiting.html")
+
+    @bp.route("/upgrade", methods=["GET", "POST"])
+    @authentication_required
+    def upgrade() -> Response | str:
+        if request.method == "GET":
```

Comment:
Why are you redirecting on `GET`? Can we just make this a `POST` only endpoint?

Reply:
Yeah, I don't know why I was doing that